File: client/trader.js
import React from 'react';


export class ArbitrageTab extends React.Component {

    constructor(props) {
        super(props);
        this.state = {data: []};
    }

    componentDidMount() {
        this._updateState(this.props);
    }

    componentWillReceiveProps(props) {
        this._updateState(props)
    }

    _updateState(props) {
        var that = this;
        $.get("/arbitrage", function (data) {
            that.setState({data: data});
        });
    }

    render() {
        if (this.state.data.length == 0) {
            return <p>No results.</p>
        }

        var rows = this.state.data.map(function (r, i) {
            return <tr key={"key-" + i}>
                <td>{r.ArbitrageId.substring(0, 10)}...</td>
                <td>{r.Date.substring(0, 16)}</td>
                <td>{r.BuyEx}</td>
                <td>{r.SellEx}</td>
                <td>{r.BuyPrice}</td>
                <td>{r.RealBuyPrice != null ? r.RealBuyPrice.toFixed(6) : '-'}</td>
                <td>{r.SellPrice}</td>
                <td>{r.RealSellPrice != null ? r.RealSellPrice.toFixed(6) : '-'}</td>
                <td>{r.Spread.toFixed(2)}%</td>
                <td>{r.RealSpread != null ? r.RealSpread.toFixed(2) + '%' : '-'}</td>
                <td>{r.Vol}</td>
                <td>{r.RealBuyVol != null ? r.RealBuyVol.toFixed(6) : '-'}</td>
                <td>{r.RealSellVol != null ? r.RealSellVol.toFixed(6) : '-'}</td>
            </tr>
        });

        return <div>
            <h1>Arbitrage</h1>
            <table className="report">
                <thead>
                    <tr>
                        <th>Id</th>
                        <th>Date</th>
                        <th>Buy<br/>Ex</th>
                        <th>Sell<br/>Ex</th>
                        <th>Buy<br/>Price</th>
                        <th>Real Buy<br/>Price</th>
                        <th>Sell<br/>Price</th>
                        <th>Real Sell<br/>Price</th>
                        <th>Margin<br/>(%)</th>
                        <th>Real<br/>Margin (%)</th>
                        <th>Vol</th>
                        <th>Buy Vol</th>
                        <th>Sell Vol</th>
                    </tr>
                </thead>
                <tbody>{rows}</tbody>
            </table>
        </div>
    }
};

export class TradeTab extends React.Component {

    constructor(props) {
        super(props);
        this.state = {data: []};
    }

    componentDidMount() {
        this._updateState(this.props);
    }

    componentWillReceiveProps(props) {
        this._updateState(props)
    }

    _updateState(props) {
        var that = this;
        $.get("/trade", function (data) {
            that.setState({data: data});
        });
    }

    render() {
        if (this.state.data.length == 0) {
            return <p>No results.</p>
        }

        var rows = this.state.data.map(function (r, i) {
            return <tr key={"key-" + i}>
                <td>{r.ArbitrageId.substring(5)}...</td>
                <td>{r.TradeId}</td>
                <td>{r.Price}</td>
                <td>{r.Quantity}</td>
                <td>{r.Pair}</td>
                <td>{r.Side}</td>
                <td>{r.Fee}</td>
                <td>{r.FeeCurrency}</td>
            </tr>
        });

        return <div>
            <h1>Trades</h1>
            <table>
                <thead>
                    <tr>
                        <th>Arbitrage Id</th>
                        <th>Trade Id</th>
                        <th>Price</th>
                        <th>Quantity</th>
                        <th>Pair</th>
                        <th>Side</th>
                        <th>Fee</th>
                        <th>Fee Currency</th>
                    </tr>
                </thead>
                <tbody>{rows}</tbody>
            </table>
        </div>
    }
};

File: src/services/web/trader.go
package main

import (
	"fmt"

	"bitbot/errorutils"

	"github.com/jmoiron/sqlx"
)

func tradedArbitrages(db *sqlx.DB, limit int) interface{} {
	const stmt = `
        select
            a.arbitrage_id,
            a.buy_ex,
            a.sell_ex,
            a.pair,
            a.ts,
            a.buy_price,
            a.sell_price,
            a.vol,
            a.spread,
            buy.price as real_buy_price,
            sell.price as real_sell_price
        from
            arbitrage a
            left join trade as buy on a.arbitrage_id = buy.arbitrage_id and buy.side = 'buy'
            left join trade as sell on a.arbitrage_id = sell.arbitrage_id and sell.side = 'sell'
        order by
            a.ts desc
        limit
            %d
    `
	var rows []*struct {
		ArbitrageId   string   `db:"arbitrage_id"`
		BuyEx         string   `db:"buy_ex"`
		SellEx        string   `db:"sell_ex"`
		Pair          string   `db:"pair"`
		Date          string   `db:"ts"`
		BuyPrice      float64  `db:"buy_price"`
		SellPrice     float64  `db:"sell_price"`
		Vol           float64  `db:"vol"`
		Spread        float64  `db:"spread"`
		RealBuyPrice  *float64 `db:"real_buy_price"`
		RealSellPrice *float64 `db:"real_sell_price"`
	}

	err := db.Select(&rows, fmt.Sprintf(stmt, limit))
	errorutils.PanicOnError(err)
	return rows
}

func trades(db *sqlx.DB, limit int) interface{} {
	const stmt = `
        select
		    arbitrage_id,
		    trade_id,
		    price,
		    quantity,
		    pair,
		    side,
		    fee,
		    fee_currency
        from
            trade
        order by
            arbitrage_id desc
        limit
            %d
    `
	var rows []*struct {
		ArbitrageId string  `db:"arbitrage_id"`
		TradeId     string  `db:"trade_id"`
		Price       float64 `db:"price"`
		Quantity    float64 `db:"quantity"`
		Pair        string  `db:"pair"`
		Side        string  `db:"side"`
		Fee         float64 `db:"fee"`
		FeeCurrency string  `db:"fee_currency"`
	}

	err := db.Select(&rows, fmt.Sprintf(stmt, limit))
	errorutils.PanicOnError(err)
	return rows
}
